### DISCUSSION TITLE
Convert ezcad2 to svg

### DISCUSSION THREAD

**Discussion by seghers** [meerk40t/meerk40t#2331] [Q&A] [answered]
The conversion is only possible on Windows with ezcad2 installed?

**Comment by tiger12506** [comment, marked as answer]
No, MeerK40t does not depend on ezcad to do this.

**Reply by seghers** [reply]
ok, thank you.

I must have a different problem:
linux
v0.9.3010

during opening/importing an .ezd I get an error.
```
File is Malformed
list index out of range 
```

**Reply by tiger12506** [reply]
Would it be possible to share the file? meerk40t devs usually hang out on Discord. Tatarize can help if you need to share it privately.

**Reply by seghers** [reply]
[ezcadfile.zip](https://github.com/meerk40t/meerk40t/files/13625396/ezcadfile.zip)

I zipped 1 file with few instructions.
I have 100+ files, I've tried 10+ but all gave the same error.

**Reply by tatarize** [reply]
Ok, I'll check it out.

**Reply by tatarize** [reply]
![image](https://github.com/meerk40t/meerk40t/assets/3302478/786d82b2-9f46-47f3-802c-3b7fe8377c14)

The Hatches in this file only goes up to 42 elements. There is no `hatch1_count`, `hatch2_count`, `hatch3_count` values.

**Reply by tatarize** [reply]
Fixed with #2334

Pushed to main.

You'd could download the source-code for main branch or if you have python and git installed:
`pip install git+https://github.com/meerk40t/meerk40t#egg=meerk40t`

**Comment by seghers** [comment]
Thank you it works

Is there a way to do this using the cli?

**Reply by seghers** [reply]
``` sh
meerk40t hello.ezd -Z -q --output hello.svg
```
This works but the cmd is never finished.

**Reply by tatarize** [reply]
If you did `-e "save hello.svg" -e "quit"` would work without needing to run anything. Both would be sent to console, and need to execute in sequence.

The quit in `-q` happens after the file is burned, and you're not really burning anything. Might need some adjustments otherwise I think.

**Reply by seghers** [reply]
it works, thank you

``` bash
    for filepath in "$@"
    do
        filepathwithoutextension="${filepath%.*}"
        meerk40t "${filepath}" --no-gui -e "save ${filepathwithoutextension}.svg" -e "quit"
    done
```

using nix or NixOS ? the following might help
``` nix
{ config
, lib
, pkgs
, ...
}:
with lib; let
  cfg = config.ncfg.programs.editors.meerk40t;

  meerk40t-ezcad-to-svg = pkgs.writeShellScriptBin "meerk40t-ezcad-to-svg" ''
    for filepath in "''$@"
    do
        filepathwithoutextension="''${filepath%.*}"
        meerk40t "''${filepath}" --no-gui -e "save ''${filepathwithoutextension}.svg" -e "quit"
    done

    # filepath=''${1}
    # filepathwithoutextension="''${filepath%.*}"
    # meerk40t "''${filepath}" --no-gui -e "save ''${filepathwithoutextension}.svg" -e "quit"
  '';
in
{
  options.ncfg.programs.editors.meerk40t.enable = lib.mkEnableOption "Enable meerk40t";

  config = lib.mkIf cfg.enable {
    home-manager.users.${config.ncfg.primaryUserName} = { ... }: {
      home.packages = with pkgs; [
        # meerk40t

        meerk40t-ezcad-to-svg

        (pkgs.meerk40t.overrideAttrs (old: {
          pname = "MeerK40t";
          # version = "0.9.3010";
          # src = pkgs.fetchFromGitHub {
          #   owner = "meerk40t";
          #   # repo = pname;
          #   repo = "MeerK40t";
          #   # rev = "refs/tags/${version}";
          #   rev = "refs/tags/0.9.3010";
          #   hash = "sha256-RlIWqxmUiL1gFMxwcdWxDiebmEzVz6kTaSlAZHr8S+I=";
          # };
          version = "0.9.4beta1";
          src = pkgs.fetchFromGitHub {
            owner = "meerk40t";
            repo = "MeerK40t";
            rev = "65ff9e5a54ede9b778b41b0adc64775447f405be"; # 20231210
            hash = "sha256-9gY41maVhZvYEugtTltgZkF4/mq+Mpdqo6mUGkpOYa0=";
          };
        }))
      ];
    };
  };
}
```

**Reply by tatarize** [reply]
I'm not too concerned. The program can load ezcad2 files because it can execute ezcad2 files on galvo laser. It wouldn't be too hard to take the `ezd.py` file and hack it to just directly output svg if that was a thing I needed. 

If you run into any other problem files give me an example of it and I'll fix up the code.

**Reply by seghers** [reply]
I'm currently converting 600 files. haven't seen an error.

The ezcad version was: 2.12.1

I won't turn the svg's into lightburn. But if I want in the future, I have a background svg to work from.

**Reply by tatarize** [reply]
Well, there are a couple things in the files that aren't 1:1 transferred over but these are mostly things like `input` and `output` operations. Like if you had sequence in there to trigger a pen-rotary. It will save in ezcad and would move to meerk40t, but it wouldn't typically save out to SVG, also things like `speed` and `power` settings, the ezcad files also have all 255 pens that were in the system at the time you saved to disk, and that doesn't even have a place in meerk40t. As a rule all the hatches contain the actual `hatch` lines that ezcad generates and same for things like spirals. and sometimes you might want to pick and choose to preserve that or not (though you can easier delete it that recreate it).

But, it would appear in the SVG file if you can read the meerk40t specific parts of the SVG that spec says all SVG readers should ignore.

**Reply by tatarize** [reply]
Also things go a lot faster if you do `pip install bitarray` since it'll use the bitarray c binding.

**Comment by seghers** [comment]
Did a test on +-20 files. 
Went from 00:01:57 to 00:01:05

in nix added bitarray to the 'propagatedBuildInputs'
``` nix
          propagatedBuildInputs = with python3.pkgs; [
            ezdxf
            meerk40t-camera
            opencv4
            pillow
            pyserial
            pyusb
            setuptools
            wxPython_4_2
            bitarray
          ];
```

**Reply by tatarize** [reply]
If very small vector files it won't have a giant impact but the bigger the file the worse that decompress huffman table lookup will get. It's generally always tolerable, but if you're planning on doing mass conversions then you might as well take the somewhat reasonable speedup.
